Add initial support for decompilers

dgileadi opened the pull request:
With the new support in vscode-java for [contributing extensions](https://github.com/redhat-developer/vscode-java/wiki/Contribute-a-Java-Extension) I thought I'd contribute an extension for decompiling class files (see #297). This implements an extension point in eclipse.jdt.ls for decompilers.

Code critiques are welcome.

I've created an initial version of a [decompiler plugin](https://github.com/dgileadi/eclipse.jdt.ls.decompiler.fernflower) and a [vscode extension](https://github.com/dgileadi/vscode-java-decompiler) as well. In the case that this PR is accepted I'll publish them at some point.

genie-ls commented:
Can one of the admins verify this patch?

fbricon commented:
ok to test Jenkins

@dgileadi that's an awesome contribution :-) I've been thinking about implementing it pretty much the same way, but lack the time to do it. It's great someone contributed such an extension.

I just skimmed over the code for now, I'll review more thoroughly in a couple days, but the overall idea looks good to me.

However you'll need to sign the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php) before we can proceed further.

fbricon commented:
@dgileadi  BTW, your decompiler plugin needs to change its namespace/packages, `org.eclipse` is reserved for projects under developed under the Eclipse.org umbrella.

fbricon commented:
@dgileadi the build  is failing: https://ci.eclipse.org/ls/job/jdt-ls-pr/331/testReport/junit/org.eclipse.jdt.ls.core.internal.handlers/

dgileadi commented:
Thanks! I just signed the ECE. I'll change the package name and check out the failures.

tsmaeder commented:
@fbricon about the package names: this is a contribution to an eclipse project under the ECA, right? Same as you do, no? Why would org.eclipse.... not be ok?

fbricon commented:
@tsmaeder org.eclipse packages in this PR are ok, since this is going into eclipse.jdt.ls. But I was referring to the actual decompiler extension: https://github.com/dgileadi/eclipse.jdt.ls.decompiler.fernflower.

@dgileadi Thanks for handling this package issue. However, your test plugin doesn't seem properly renamed. You should probably rename your git repo as well.
Can you please sign your commits as well? This is required by the Eclipse foundation for all external contributions. See https://dev.eclipse.org/eclipse-webhook/services/status_details.php?id=59c6e531b061c


Jenkins build got stuck for this PR, I'll kick it again.

dgileadi commented:
I renamed the plugin (hopefully properly this time). I also rewrote the history to use the correct email address. Hopefully that'll allow it to pass validation.

fbricon commented:
@dgileadi build is fixed but IP Validation still fails. I suggest you squash all your commits, [use the Signed-Off-By](https://dev.eclipse.org/eclipse-webhook/services/status_details.php?id=59c9199725be9), which must match the email address you signed the ECA under.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandler.java:
```diff
@@ -0,0 +1,166 @@
+
+public class DecompileCommandHandler {
+
+	private static final String EXTENSION_POINT_ID = "org.eclipse.jdt.ls.core.decompilerCommandHandler";
```
EXTENSION_POINT_ID = "org.eclipse.jdt.ls.core.decompiler"

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandler.java:
```diff
@@ -0,0 +1,166 @@
+			decompilerCommandHandler = null;
+		}
+
+		public synchronized IDecompilerCommandHandler getDecompilerCommandHandler() {
```
getDecompiler()

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandler.java:
```diff
@@ -0,0 +1,166 @@
+					if (extension instanceof IDecompilerCommandHandler) {
+						decompilerCommandHandler = (IDecompilerCommandHandler) extension;
+					} else {
+						String message = "Invalid extension to " + EXTENSION_POINT_ID + ". Must implement org.eclipse.jdt.ls.core.internal.IDecompilerCommandHandler";
```
... must implement IDecompiler

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandler.java:
```diff
@@ -0,0 +1,166 @@
+						return null;
+					}
+				} catch (CoreException e) {
+					JavaLanguageServerPlugin.logException("Unable to create decompiler command handler ", e);
```
Unable to create decompiler

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandler.java:
```diff
@@ -0,0 +1,166 @@
+		}
+
+		Set<DecompilerCommandHandlerDescriptor> handlers = getDecompilerCommandHandlerDescriptors();
+		if (handlers.isEmpty()) {
```
should fall back on default decompiler (i.e current disassembler)

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java:
```diff
@@ -127,6 +137,8 @@
 	private boolean renameEnabled;
 	private MemberSortOrder memberOrders;
+	private String decompilerId;
+	private String decompilerConfiguration;
```
decompilerConfiguration should be a `Map<String,Object>`.
eg.
```
java.decompiler.configuration = {
   fernflower {
       foo:'bar'
   },
   jad {
     ...
   }
}
```

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/ClassfileContentHandlerTest.java:
```diff
@@ -30,20 +34,26 @@
-		handler = new ClassfileContentHandler();
+		preferenceManager = mock(PreferenceManager.class);
+		preferences = mock(Preferences.class);
+		when(preferences.getDecompilerId()).thenReturn("none");
```
by default, it should return "default" (i.e disassembler). "none" implies no disassembly/decompilation will be attempted whatsoever (Which could be a possibility)

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandlerTest.java:
```diff
@@ -0,0 +1,82 @@
+	public void testDecompileUnknownDecompiler() {
+		when(preferences.getDecompilerId()).thenReturn("noDecompiler");
+
+		String result = handler.decompile(classFile, monitor);
```
should still default to standard dissasembly

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/DecompileCommandHandlerTest.java:
```diff
@@ -0,0 +1,82 @@
+		when(preferences.getDecompilerId()).thenReturn("placeholderDecompiler");
+
+		String result = handler.decompile(classFile, monitor);
+		assertNull(result);
```
should still default to standard dissasembly

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/IDecompilerCommandHandler.java:
```diff
@@ -0,0 +1,33 @@
+	 * @return decompiled bytecode or <code>null</code>
+	 * @throws Exception
+	 */
+	String decompile(IClassFile classFile, String configuration, IProgressMonitor monitor) throws Exception;
```
configuration should be a `Map<String, Object>`

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/NavigateToDefinitionHandlerTest.java:
```diff
@@ -64,6 +73,7 @@ public void testDisassembledSource() throws Exception {
 	}
 
 	private void testClass(String className, int line, int column) throws JavaModelException {
+		when(preferences.getDecompilerId()).thenReturn("none");
```
thenReturn "default"

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/IDecompiler.java:
```diff
@@ -0,0 +1,36 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
```
copyright David Gileadi

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/Disassembler.java:
```diff
@@ -0,0 +1,39 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
```
David Gileadi - default IDecompiler implementation

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ClassfileContentHandler.java:
```diff
@@ -16,36 +16,60 @@
 		} catch (JavaModelException e) {
 			JavaLanguageServerPlugin.logException("Exception getting java element ", e);
 		}
-		if (source == null) {
```
either this null-check needs to stay, or decompile(...) needs to guarantee non-null values (eg. in case of exception)

dgileadi replied:
Not so—this method is called in two places:

1. By the `contents` method above, which has the original null check
2. By `JDTUtils.findElementsAtSelection`, which requires a null if no content was found

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java:
```diff
@@ -0,0 +1,192 @@
+		}
+
+		public synchronized IDecompiler getDecompiler() {
+			if (decompiler == null) {
```
I'm not sure the decompiler should be cached. That'd force extensions to make their decompilers as singletons. I don't want to make any assumption at that point. 
Unless there are performance concerns (which will need to be proven), there's no need to keep a decompiler instance in memory

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java:
```diff
@@ -0,0 +1,192 @@
+
+		String decompilerId = preferenceManager.getPreferences().getDecompilerId();
+		Optional<DecompilerDescriptor> handler = handlers.stream().filter(d -> decompilerId == null || decompilerId.equals(d.id)).findFirst();
+		if (!handler.isPresent()) {
```
if there's no matching decompiler, we should log it and fall back on the disassembler

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java:
```diff
@@ -228,6 +243,32 @@ private static String getStringValue(Map<String, Object> configuration, String k
 		return def;
 	}
 
+	private static Map<String, Object> getMapValue(Map<String, Object> configuration, String key, Map<String, Object> def) {
```
can we have a unit test that checks decompiler config is properly parsed as a Map?

fbricon replied:
Now that I think of it, incoming maps are flattened via MapFlattener.flatten in JDTLanguageServer.didChangeConfiguration, so that might be a problem eventually. Have you tried the fernflower config with this change?

dgileadi replied:
I haven't, but if maps are being flattened I can't see how it would work. Should I switch back to my command-line-arg settings, or what do you suggest?

fbricon replied:
I can think of many ugly ways to do it, but I need to sleep on it. I need to finish some other task today before I can focus on that issue.

fbricon replied:
Ok so I'm gonna change how flattening is done to make it compatible with nested maps. I got a bit if refactoring to do.

dgileadi replied:
Excellent.

fbricon replied:
See #370

fbricon replied:
@dgileadi #370 is merged. Can you please rebase against master, see if it works for you?

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java:
```diff
@@ -288,6 +329,12 @@ public static Preferences createFrom(Map<String, Object> configuration) {
 		String favoriteStaticMembers = getStringValue(configuration, FAVORITE_STATIC_MEMBERS, "");
 		prefs.setFavoriteStaticMembers(favoriteStaticMembers);
 
+		String decompilerId = getStringValue(configuration, DECOMPILER_ID_KEY, null);
```
I'm wondering wether we should check here the decompilerId is exists, and if not, make sure the disassembler one is used instead

dgileadi replied:
I don't think so—I don't think Preferences should handle that logic, and DecompilerManager now does that internally.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java:
```diff
@@ -0,0 +1,192 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
```
Copyright (c) 2017 David Gileadi and others.

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/TestDecompiler.java:
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
```
Copyright (c) 2017 David Gileadi and others.

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManagerTest.java:
```diff
@@ -0,0 +1,86 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
```
Copyright (c) 2017 David Gileadi and others.

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManagerTest.java:
```diff
@@ -0,0 +1,86 @@
+		when(preferences.getDecompilerId()).thenReturn("noDecompiler");
+
+		String result = handler.decompile(classFile, monitor);
+		assertNull(result);
```
when it's an unknown decompiler it should fall back to the disassembler

fbricon commented on org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManagerTest.java:
```diff
@@ -0,0 +1,86 @@
+		when(preferences.getDecompilerId()).thenReturn("placeholderDecompiler");
+
+		String result = handler.decompile(classFile, monitor);
+		assertNull(result);
```
when it's an non-existing decompiler it should fall back to the disassembler

dgileadi commented:
Verified that the changes in #370 work for passing decompiler preferences from vscode. Also nice refactor—I like the pull-style API.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java:
```diff
@@ -0,0 +1,186 @@
+	private String prependHeader(String decompiledCode, IDecompiler decompiler) {
+		if (decompiledCode != null) {
+			String header = String.format(DECOMPILED_HEADER, getDecompilerName());
+			if (decompiler instanceof Disassembler) {
```
Actually the Disassembler class should add that comment.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java:
```diff
@@ -0,0 +1,186 @@
+public class DecompilerManager {
+
+	public static final String DECOMPILED_HEADER = " // Failed to get sources. Instead, stub sources have been generated by the %s.";
+	public static final String DISASSEMBLER_HEADER_ADDITION = " // Implementation of methods is unavailable.\n";
```
this comment belongs to the disassembler

dgileadi replied:
I waffled between the different ways and had decided that the job of adding messages belonged to DecompilerManager. However I'm plenty happy to switch it.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/Disassembler.java:
```diff
@@ -0,0 +1,40 @@
+		ClassFileBytesDisassembler disassembler = ToolFactory.createDefaultClassFileBytesDisassembler();
+		String disassembledByteCode = null;
+		try {
+			disassembledByteCode = disassembler.disassemble(classFile.getBytes(), LF, ClassFileBytesDisassembler.WORKING_COPY);
```
should prepend comment about empty methods here

fbricon commented:
Ok, so I after a lengthy discussion with @gorkem, we've come to the conclusion we need to make the solution a bit more generic, but that'll require more work.

So basically, the idea is to treat decompilers as content providers (IContentProvider). Content providers would be contributed to the server through extension points. When content is requested for a certain URI, we'd loop through all content providers, find the ones matching that URI, then order them by priority. If 1st provider fails to generate content, then fall back to the next and so on.

Any content provider with priority 0, will be called 1st and can not be overridden. 

So considering the extension point:
```
contentProvider {
   -id: String, required provider unique id
   -label: String, optional label
   -class: Class implementing IContentProvider/Factory returning IContentProvider, required
   -priority: 0 or positive int. Lowest value has highest priority. "0" priority means the content provider will always be called 1st, even if user defines a preferred provider
   -uriPattern: regexp that matches URIs this contentProvider can support.
   -cacheable: optional boolean, indicates whether the server can safely cache content for a period of time, true by default.
} 
```
So, we'd have by default 2 contentProviders:
```
contentProvider {
 -id:sourceContentProvider
 -class:SourceContentProvider (reads from source jars)
 -priority:0
 -uriPattern: "jdt://contents/.*\.class.*"
 -cacheable:false 
 }
contentProvider {
 -id:disassemblerContentProvider
 -class:DisassemblerContentProvider (current byte code disassembler)
 -priority:1000
 -uriPattern: "jdt://contents/.*\.class.*"
 -cacheable:true 
 }
```
Then a new contentProvider could be contributed by a 3rd party extension, that would be called after reading sources but before the standard disassembler:
```
contentProvider {
 -id:fernflowerContentProvider
 -class:FernflowerContentProvider (current byte code disassembler)
 -priority:500
 -uriPattern: "jdt://contents/.*\.class.*"
 -cacheable:true 
 }
```
The new IContentProvider interface could look like
IContentProvider {
   void setPreferences(Preferences preferences)
   String getContent(URI uri, monitor);
}

The ContentProviderManager (ex-DecompilerManager)  would receive a query to get content for a given uri. 1st, it needs to collect all content providers that match that uri (via uriPattern). The remaining providers would then be sorted. The default order would be based on the priority set per content provider. 
So if there are competing content providers for instance (eg: fernflower(priority:500), otherdecompiler(priority:400)) , and the user wants to override their priority, he'd add this setting: 
`java.contentProvider.preferred: "fernflower, otherdecompiler"` 
No matter what, the content provider with priority 0 would still be called.  fernflower would be given priority 1, otherdecompiler would receive priority 2.

ContentProviderManager would not inject any headers after calling the content providers. Each providers would be responsible for injecting source header if necessary.

As for configuration support, we should have a Preferences.asMap() that'll return the original server configuration Map (that should simplify things), should content provider need access to settings beyond their own, like java.home.

WDYT?

dgileadi commented:
First of all, I like the idea of making things generic—it seems useful for additional things in the future. However before running with the generic approach I'd be interested in examples of what other types of content might be. They'd have to be things that can be converted to String, and which aren't easily provided by simply resolving the contents of a URL (since the existing code does that just fine). I haven't thought of any, but I've been away from the Java world for a little while and may simply be missing some obvious things.

I'm unsure about the priority scheme for choosing a content provider. On the authoring side, it would be hard to pick a priority—"Should I be fair and use the same priority as Fernflower? What is Fernflower's priority anyway?" Or "My decompiler is worse but I really want to market it so I'm giving it priority 1! Take that!"

On the user side it's also a bit confusing—for decompilers I really only want to chose one provider. `java.contentProvider.preferred: "fernflower, otherdecompiler"` doesn't make sense, because I don't expect a decompiler to fail to return content and thus fall back on a second one. I just want "fernflower" or "procyon" or whatever.

It's also confusing if there are future content providers for other things (whatever they may be): say I have `java.contentProvider.preferred: "fernflower, gadgetprovider"`. In this case the two providers are mutually exclusive, so the ordering doesn't matter anyway.

Regardless, I think we first need to think of some examples of other kinds of content providers and see how they might be used/configured, and that might help determine how best to create a generic content provider extension point.

gorkem commented:
Examples off the top of my head, we may want to have a content provider that treats the module-info.class on a jar file differently than simply decompiling it. Or a content provider to list the contents of a .jar file.

fbricon commented:
```
I'm unsure about the priority scheme for choosing a content provider. On the authoring side, it would be hard to pick a priority—"Should I be fair and use the same priority as Fernflower? What is Fernflower's priority anyway?" Or "My decompiler is worse but I really want to market it so I'm giving it priority 1! Take that!"
```
Maybe we can let vscode-java detect a new compiler has been added and, in case several decompilers are available, ask the user which one is preferred. 
The priority bit is mostly important so that we can ensure the source provider goes 1st, the disassembler goes last as a fallback.

```
On the user side it's also a bit confusing—for decompilers I really only want to chose one provider. `java.contentProvider.preferred: "fernflower, otherdecompiler"` doesn't make sense, because I don't expect a decompiler to fail to return content and thus fall back on a second one. I just want "fernflower" or "procyon" or whatever.
```
Most likely, users will just add `java.contentProvider.preferred: "fernflower"`, in case there are multiple choices. If there's only one, it'll be used by default, no need to change any settings. I'd say the fall back mechanism is a side effect of the implementation, but it's not a bad one, so I'm fine with it.


```
It's also confusing if there are future content providers for other things (whatever they may be): say I have `java.contentProvider.preferred: "fernflower, gadgetprovider"`. In this case the two providers are mutually exclusive, so the ordering doesn't matter anyway.
```
Indeed, multiple non competing content providers can be set, in that case, the order will have no impact on one another. Also I don't expect there will be a plethora of content provider TBH :-)

```
Regardless, I think we first need to think of some examples of other kinds of content providers and see how they might be used/configured, and that might help determine how best to create a generic content provider extension point.
```
See @gorkem's answer. We also thought about having a special provider to deal with jsp classes, if we ever have to deal with that.

dgileadi commented:
I've been thinking about this. Regarding the examples, I'd really hope that a jar file would be treated as a browsable directory. Having it simply produce a text list of its contents would be a fail, IMO.

The other two examples are of specialized decompilation, so they could fairly be called decompilers. The only example of a content provider that isn't a decompiler is the code that searches for source code for a class file. Currently the only such provider is internal, although I can imagine creating one that scours the 'net for sources and downloads them.

In any case my issue with the proposal is still the numeric priority, because the only place it makes sense is to the internal code (to make sure we try loading source first and disassembling last). For extension authors and vscode users it's just confusing.

I like your idea of having vscode-java detect multiple external content providers and ask the user to choose a preferred one. I don't know how this would work in code, but it's a good solution to the ordering problem.

I'd prefer to keep any priority numbering internal to jdt-ls.

fbricon commented:
Would it be ok to make the priority optional? It'd still appear in the xsd but adopters wouldn't have to worry about it

dgileadi commented:
Yes, making it optional would work.

When multiple providers with the same priority are detected, how should jdt-ls communicate that to vscode-java? Just log an error in the "Language Support for Java" output, something like the following?

    "You have more than one content provider installed: ["fernflower", "procyon"]. Please use the "java.contentProvider.preferred" setting to choose which one you want to use."

Or is there a built-in way to request information via a prompt?

fbricon commented:
I'd start by logging a message, later we can iterate over how to handle that on the client side.

fbricon commented:
As for returning content from a jar as a tree of whatever, that means IContentProvider.getContent would have to return an Object.

dgileadi commented:
Yes, if it's implemented as an IContentProvider. I tend to think that browsing a jar file should be handled natively by vscode (either by the -java extension or some other extension) and the role of jdt-ls should be to return the content of clicked-on .class files within the jar.

fbricon commented:
I'll leave that responsability to an other extension for now :-)

dgileadi commented:
Question: why separate methods for `setPreferences` and `getContent`? Why not just pass the preferences as a parameter to `getContent`?

fbricon commented:
I tend to favour the separation of concerns. Passing those preferences all across those methods feel weird, out of place.

dgileadi commented:
So basically just a personal preference :) I'm fine with that.

The side effect is that it makes content provider classes need to hold state, which theoretically could affect multithreading some day if we cache provider classes, but I don't think that's an actual concern.

fbricon commented:
indeed, it's a personal preference, but you raise a valid point about caching. 
I guess I'm fine with either way now :-)

dgileadi commented:
Okay, a bigger issue: it's fairly simple to go from a URI to an IClassFile, thanks to JDTUtils. However it doesn't seem possible to go the other direction. We need to go the other direction because handlers like NavigateToDefinitionHandler call `JDTUtils.findElementsAtSelection`, passing a `unit` of type IClassFile (and no URI). `findElementsAtSelection` loads the source code by any means possible, including decompiling/disassembling it. However with the current IContentProvider design it can't do this without a URI.

I'd hoped that the class file's `getReference` would return non-null, allowing `JDTUtils.getFileURI(IReference)` to provide the URI, but no such luck.

My preferred approach would be to limit IContentProvider to providing source for just IClassFile instances, instead of URIs. That way we can just access the content of the class file directly. Barring that, what do you suggest?

dgileadi commented:
Any ideas? Shall I just switch to only supporting IClassFile?

fbricon commented:
@dgileadi  in that case we might actually need to keep an IDecompiler interface around. So if we have the IContentProvider extension and the actual implem is also an instance of IDecompiler, then we could call something like `((IDecompiler) provider).getContent(classFile, monitor)` method.

dgileadi commented:
So for cases like NavigateToDefinitionHandler where we only have an IClassFile, if the extension weren't an IDecompiler then it wouldn't be considered as a candidate to provide source? I can see how this would work, but I don't like it.

Again, what's an example of something that we'd provide text content for that isn't a class file? The only prior example I recall was listing the contents of a .jar file, which to me seems like a bad idea in the first place (and doesn't seem to justify the additional IContentProvider/IDecompiler complexity). Any others?

fbricon commented:
If you open a single class file that's not part of a project, then calling the decompiler from a URI makes sense

dgileadi commented:
That does make sense.

Is there a good reason to have two IDecompiler/IContentProvider interfaces, versus just a single interface with two overloads of `getContent` (or `decompile`)?

> On Oct 10, 2017, at 2:41 PM, Fred Bricon <notifications@github.com> wrote:
> 
> If you open a single class file that's not part of a project, then calling the decompiler from a URI makes sense
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse/eclipse.jdt.ls/pull/360#issuecomment-335616715>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAViPh-xb_GoMBZQMA_CuiNjNzTOZ2L2ks5sq-RtgaJpZM4Phq1m>.
>

fbricon commented:
IContentProvider is not necessarily a decompiler in @gorkem's mind. So I'd rather keep them separate, personally.

fbricon commented on org.eclipse.jdt.ls.core/plugin.xml:
```diff
@@ -29,4 +30,23 @@
    </extension>
+   <extension
+         point="org.eclipse.jdt.ls.core.contentProvider">
+      <decompiler
```
should be `<contentProvider`

fbricon commented on org.eclipse.jdt.ls.core/plugin.xml:
```diff
@@ -29,4 +30,23 @@
+   </extension>
+   <extension
+         point="org.eclipse.jdt.ls.core.contentProvider">
+      <decompiler
```
should be `<contentProvider`

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/IDecompiler.java:
```diff
@@ -0,0 +1,32 @@
+	 * @return text content or <code>null</code>
+	 * @throws CoreException
+	 */
+	public String decompile(IClassFile classFile, IProgressMonitor monitor) throws CoreException;
```
I don't like decompile because SourceContentProvider doesn't decompile anything, and IDecompiler.getContent(classfile) wouldn't feel very natural either. How about getSource(classfile)?

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ContentProviderManager.java:
```diff
@@ -0,0 +1,237 @@
+	 * @return the text content if successful, or <code>null</code> if unsuccessful,
+	 *         or an empty string if canceled
+	 */
+	public String decompile(IClassFile classFile, IProgressMonitor monitor) {
```
getSource

fbricon commented:
That looks pretty good so far, but I'll give a more detailed look tomorrow. 
So currently the contribution exceeds 1000LOC, so we'll need to open a [CQ](https://www.eclipse.org/projects/dev_process/ip-process-in-cartoons.php), that will have to be approved by the Eclipse Legal team before we can merge it, which will take a few days.

But, I believe we no longer need the getMap() methods in https://github.com/eclipse/eclipse.jdt.ls/pull/360/files#diff-6d9005c681b8bd4f0707f2ef68e36515R87, so if removing that and the associated test code makes the contribution magically shrink to under 1000LOC, we'll be good :-)

fbricon commented:
Also, can you please rebase against master?

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/IContentProvider.java:
```diff
@@ -0,0 +1,43 @@
+	 *
+	 * @param preferences
+	 */
+	void setPreferences(Preferences preferences);
```
`default void setPreferences(Preferences preferences){}` so you don't need to implement empty methods

gorkem commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/preferences/Preferences.java:
```diff
@@ -105,6 +107,17 @@
+	/**
+	 * Preference key for the maximum size of the content cache.
+	 */
+	public static final String CONTENT_PROVIDER_CACHE_MAX_SIZE_KEY = "java.contentProvider.cache.maxSize";
```
Does this need to be a user configurable preference? I think we should just remove it as a preference and have a constant somewhere for it.

dgileadi replied:
I'll make that change.

dgileadi commented on org.eclipse.jdt.ls.core/schema/org.eclipse.jdt.ls.core.contentProvider.exsd:
```diff
@@ -0,0 +1,115 @@
+         <attribute name="class" type="string" use="required">
+            <annotation>
+               <documentation>
+                  Name of a class that implements IContentProvider. This class may optionaly implement IDecompiler instead.
```
I think we do—IContentProviders that don't also implement IDecompiler won't be able to provide source for things like NavigateToDefinitionHandler, for example. So we want to encourage implementing IDecompiler, and that means documenting it.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java:
```diff
@@ -466,7 +463,8 @@ public static IJavaElement findElementAtSelection(ITypeRoot unit, int line, int
 		if (unit instanceof IClassFile) {
 			IClassFile classFile = (IClassFile) unit;
-			String contents = disassemble(classFile);
+			ContentProviderManager contentProvider = new ContentProviderManager(preferenceManager);
```
We want to keep a single instance of ContentProviderManager across the JVM, or else it's cache of decompiled classes will have no value

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ContentProviderManager.java:
```diff
@@ -0,0 +1,235 @@
+
+	private final PreferenceManager preferenceManager;
+
+	private static Cache<String, String> cachedContent = CacheBuilder.newBuilder().maximumSize(MAX_CACHE_SIZE).build();
```
I'd rather avoid keeping a static cache instance. You get pretty much the same results if a single instance of ContentProviderManager is shared accross all classes. The reason static cache is bad: it introduces side effects when running unit tests, the cache would be hot, depending on the order of test runs causing potentially hard to debug issues.

You can store the ContentProviderManager instance in JavaLanguageServerPlugin along the rest of the other global "managers". (I don't quite like that pattern and would much rather prefer a dependency injection mechanism, but that's another debate)

fbricon replied:
I'm not overly familiar with Guava's cache. How does it behave during concurrent accesses?

dgileadi replied:
[Their docs](https://github.com/google/guava/wiki/CachesExplained) describe it as "similar to `ConcurrentMap`" but with the addition of features like eviction. The docs describe using locks to manage concurrent access.

fbricon commented:
And...merged! 
Thanks a lot for your contribution @dgileadi (and your patience!). 

A new version of vscode-java embedding this change will be available on Oct. 15th. You can plan to release your extension by then.

dgileadi commented:
Fantastic. Thanks for your patience with walking this through.

> On Oct 12, 2017, at 7:15 AM, Fred Bricon <notifications@github.com> wrote:
> 
> And...merged!
> Thanks a lot for your contribution @dgileadi <https://github.com/dgileadi> (and your patience!).
> 
> A new version of vscode-java embedding this change will be available on Oct. 15th. You can plan to release your extension by then.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse/eclipse.jdt.ls/pull/360#issuecomment-336150596>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAViPnBEtb-Id2sKwaLVhnMiGVRFQM79ks5srh72gaJpZM4Phq1m>.
>

fbricon commented:
@dgileadi let me know when you have an updated vscode extension I can play with. 
Also feel free to join us on https://mattermost.eclipse.org/eclipse/channels/eclipsejdtls

dgileadi commented:
I went ahead and pushed a new version.

Is there a way to detect when the preferences have changed? I think it would be a good idea to clear the decompiled source code cache whenever the decompiler preferences change, to allow the new options to take effect.

> On Oct 12, 2017, at 8:34 AM, Fred Bricon <notifications@github.com> wrote:
> 
> @dgileadi <https://github.com/dgileadi> let me know when you have an updated vscode extension I can play with.
> Also feel free to join us on https://mattermost.eclipse.org/eclipse/channels/eclipsejdtls <https://mattermost.eclipse.org/eclipse/channels/eclipsejdtls>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse/eclipse.jdt.ls/pull/360#issuecomment-336175523>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAViPngfbqdAbYURmwAmTdw9WMdVQjBnks5srjGbgaJpZM4Phq1m>.
>
